File: frontend/src/app/core/services/banner.service.ts
import { Injectable } from '@angular/core';
import { HttpClient, HttpParams } from '@angular/common/http';
import { Observable } from 'rxjs';
import { environment } from '../../../environments/environment';
import { ApiResponse, PageResponse } from '../models/api-response.model';
import {
  BannerResponse,
  BannerStatus,
  CreateBannerRequest,
  UpdateBannerRequest
} from '../models/banner.model';

@Injectable({ providedIn: 'root' })
export class BannerService {
  private apiUrl = `${environment.apiUrl}/admin/banners`;

  constructor(private http: HttpClient) {}

  list(opts: { status?: BannerStatus; position?: string; page?: number; size?: number } = {})
    : Observable<ApiResponse<PageResponse<BannerResponse>>> {
    let params = new HttpParams()
      .set('page', (opts.page ?? 0).toString())
      .set('size', (opts.size ?? 10).toString());
    if (opts.status) params = params.set('status', opts.status);
    if (opts.position) params = params.set('position', opts.position);
    return this.http.get<ApiResponse<PageResponse<BannerResponse>>>(this.apiUrl, { params });
  }

  getById(id: number): Observable<ApiResponse<BannerResponse>> {
    return this.http.get<ApiResponse<BannerResponse>>(`${this.apiUrl}/${id}`);
  }

  create(request: CreateBannerRequest): Observable<ApiResponse<BannerResponse>> {
    return this.http.post<ApiResponse<BannerResponse>>(this.apiUrl, request);
  }

  update(id: number, request: UpdateBannerRequest): Observable<ApiResponse<BannerResponse>> {
    return this.http.put<ApiResponse<BannerResponse>>(`${this.apiUrl}/${id}`, request);
  }

  publish(id: number): Observable<ApiResponse<BannerResponse>> {
    return this.http.post<ApiResponse<BannerResponse>>(`${this.apiUrl}/${id}/publish`, {});
  }

  unpublish(id: number): Observable<ApiResponse<BannerResponse>> {
    return this.http.post<ApiResponse<BannerResponse>>(`${this.apiUrl}/${id}/unpublish`, {});
  }

  delete(id: number): Observable<void> {
    return this.http.delete<void>(`${this.apiUrl}/${id}`);
  }
}

File: backend/src/main/java/com/training/marketplace/service/BannerService.java
package com.training.marketplace.service;

import com.training.marketplace.common.PageResponse;
import com.training.marketplace.dto.request.CreateBannerRequest;
import com.training.marketplace.dto.request.UpdateBannerRequest;
import com.training.marketplace.dto.response.BannerResponse;
import com.training.marketplace.enums.PublishStatus;
import org.springframework.data.domain.Pageable;

import java.util.List;

public interface BannerService {

    // --- Admin ---
    BannerResponse create(CreateBannerRequest request);

    BannerResponse update(Long id, UpdateBannerRequest request);

    BannerResponse getById(Long id);

    PageResponse<BannerResponse> list(PublishStatus status, String position, Pageable pageable);

    void delete(Long id); // soft delete (active=false)

    BannerResponse publish(Long id);

    BannerResponse unpublish(Long id);

    // --- Storefront ---
    /** Banners effective right now in a slot (PUBLISHED + within schedule), ordered (B-405). */
    List<BannerResponse> listEffectiveByPosition(String position);
}
